src/Microsoft.DocAsCode.Build.Common/ModelAttributeHandlers/ReflectionHelper.cs

```diff
@@ -36,89 +40,100 @@ public static List<PropertyInfo> GetGettableProperties(Type type)
+            return _isDictionaryCache.GetOrAdd(type, t => IsDictionaryTypeCore(t));
+        }
+
+        private static bool IsDictionaryTypeCore(Type type)
         {
-            if (typeof(IDictionary).IsAssignableFrom(t))
+            if (typeof(IDictionary).IsAssignableFrom(type))
             {
                 return true;
             }
 
-            return ImplementsGenericDefintion(t, typeof(IDictionary<,>))
-            || ImplementsGenericDefintion(t, typeof(IReadOnlyDictionary<,>));
+            return ImplementsGenericDefintion(type, typeof(IDictionary<,>)) ||
+                ImplementsGenericDefintion(type, typeof(IReadOnlyDictionary<,>));
         }
 
         public static bool IsIEnumerableType(Type t)
         {
             return typeof(IEnumerable).IsAssignableFrom(t);
         }
 
-        public static bool TryGetGenericType(Type type, Type genericTypeDefinition, out Type genericType)
+        public static Type GetGenericType(Type type, Type genericTypeDefinition)
         {
-            genericType = null;
-            if (IsGenericType(type, genericTypeDefinition))
+            return _genericTypeCache.GetOrAdd(Tuple.Create(type, genericTypeDefinition), t => GetGenericTypeNoCache(t.Item1, t.Item2));
+        }
+
+        private static Type GetGenericTypeNoCache(Type type, Type genericTypeDefinition)
+        {
+            if (type.IsInterface == genericTypeDefinition.IsInterface &&
+                IsGenericType(type, genericTypeDefinition))
             {
-                genericType = type;
-                return true;
+                return type;
             }
-            foreach (var i in type.GetInterfaces())
+            if (genericTypeDefinition.IsInterface)
             {
-                if (IsGenericType(i, genericTypeDefinition))
+                foreach (var i in type.GetInterfaces())
                 {
-                    genericType = i;
-                    return true;
+                    if (IsGenericType(i, genericTypeDefinition))
+                    {
+                        return i;
+                    }
                 }
             }
+            return null;
+        }
 
-            return false;
+        public static bool TryGetGenericType(Type type, Type genericTypeDefinition, out Type genericType)
+        {
+            genericType = GetGenericType(type, genericTypeDefinition);
+            return genericType != null;
         }
 
         public static bool ImplementsGenericDefintion(Type type, Type genericTypeDefinition)
         {
-            if (IsGenericType(type, genericTypeDefinition))
-            {
-                return true;
-            }
-
-            foreach (var i in type.GetInterfaces())
-            {
-                if (IsGenericType(i, genericTypeDefinition))
-                {
-                    return true;
-                }
-            }
-
-            return false;
+            return GetGenericType(type, genericTypeDefinition) != null;
```

Comment:
I want to remove `TryGetGenericType`.

---
In reply to: [96348446](https://github.com/dotnet/docfx/pull/1187#discussion_r96348446) [](ancestors = 96348446)